File: solutions/misc/frequently-asked-questions/word-ladder.js
/**
 * Word Ladder (LC 127) — Hard
 * Pattern: BFS on implicit graph (word → neighbors by 1-char change)
 *
 * Time:  O(M² × N) where M = word length, N = wordList size
 * Space: O(M² × N) for adjacency preprocessing
 *
 * 60-Second ID:
 *   Core DS: word → graph node; queue for BFS
 *   Core Op: shortest path (BFS guarantees this)
 *   Constraint: single character changes define edges
 *   → BFS on word graph; wildcard pattern preprocessing speeds it up
 */

/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @returns {number} length of shortest transformation sequence, or 0
 */
function ladderLength(beginWord, endWord, wordList) {
  const wordSet = new Set(wordList);
  if (!wordSet.has(endWord)) return 0;

  // Preprocessing: map pattern → list of words matching it
  // e.g., "h*t" → ["hot","hit"]
  const patternMap = new Map();
  const allWords = [beginWord, ...wordList];
  for (const word of allWords) {
    for (let i = 0; i < word.length; i++) {
      const pattern = word.slice(0, i) + "*" + word.slice(i + 1);
      if (!patternMap.has(pattern)) patternMap.set(pattern, []);
      patternMap.get(pattern).push(word);
    }
  }

  const queue = [[beginWord, 1]]; // [word, steps]
  const visited = new Set([beginWord]);

  while (queue.length > 0) {
    const [word, steps] = queue.shift();

    for (let i = 0; i < word.length; i++) {
      const pattern = word.slice(0, i) + "*" + word.slice(i + 1);
      for (const neighbor of (patternMap.get(pattern) || [])) {
        if (neighbor === endWord) return steps + 1;
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push([neighbor, steps + 1]);
        }
      }
    }
  }
  return 0;
}

// ─── Tests ───────────────────────────────────────────────────────────────────
console.log(ladderLength("hit", "cog", ["hot","dot","dog","lot","log","cog"])); // 5
console.log(ladderLength("hit", "cog", ["hot","dot","dog","lot","log"]));       // 0

module.exports = { ladderLength };

File: solutions/misc/frequently-asked-questions/word-ladder.py
"""
Word Ladder (LC 127) — Hard
Pattern: BFS on implicit graph

Time:  O(M² × N) where M = word length, N = wordList size
Space: O(M² × N)
"""

from collections import defaultdict, deque
from typing import List


def ladder_length(begin_word: str, end_word: str, word_list: List[str]) -> int:
    # Convert word_list to set for O(1) lookup
    word_set = set(word_list)
    # If end_word not in word_list, transformation is impossible
    if end_word not in word_set:
        return 0

    L = len(begin_word)

    # Preprocessing: build adjacency map using wildcard patterns
    # For each word, generate patterns like "h*t" and map to all words matching that pattern
    # This allows O(1) neighbor lookup during BFS instead of O(N) per word
    pattern_map: defaultdict = defaultdict(list)
    for word in [begin_word] + word_list:
        for i in range(L):
            pattern = word[:i] + "*" + word[i+1:]
            pattern_map[pattern].append(word)

    # BFS queue stores (current_word, number_of_steps_to_reach_it)
    queue = deque([(begin_word, 1)])
    visited = {begin_word}  # Track visited words to avoid cycles

    while queue:
        word, steps = queue.popleft()
        # Generate all possible patterns for current word
        for i in range(L):
            pattern = word[:i] + "*" + word[i+1:]
            # Explore all neighbors (words that differ by exactly 1 character)
            for neighbor in pattern_map[pattern]:
                if neighbor == end_word:
                    return steps + 1  # Found target, return path length
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((neighbor, steps + 1))

    return 0  # No transformation path exists


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print(ladder_length("hit", "cog", ["hot","dot","dog","lot","log","cog"]))  # 5
    print(ladder_length("hit", "cog", ["hot","dot","dog","lot","log"]))        # 0
    print(ladder_length("a", "c",    ["a","b","c"]))                           # 2
